### PR TITLE
Update StormScope model package

### PR DESCRIPTION
**Pull request by pzharrington**
# Earth2Studio Pull Request

## Description
Updates the StormScope model in Earth2Studio to target the refreshed`nvidia/stormscope-goes-mrms` package (commit `62f0fd2`), replacing the legacy `DiT`-wrapping approach with direct `physicsnemo.Module` loading.
 New models enable GOES, MRMS, and GLM Lightning predictions at CONUS-wide 3km / 10 minute spatiotemporal resolution. Example output:
image

## Summary

- **Drop legacy shim layer** — deleted `earth2studio/models/nn/stormscope_util.py` (`EDMPrecond`, `DropInDiT`, `model_wrap`). Checkpoints now load via `physicsnemo.Module.from_checkpoint`; the returned module is stored directly in `model_spec`. Sampler calls updated to the new preconditioner signature (`forward(x, t, condition=None)`, no `round_sigma`, no `class_labels`).

- **Registry-driven model loading** — `load_model` resolves variant names (and aliases) from `registry.json` rather than hard-coded indices. Normalization statistics and `variables`/`conditioning_vars` lists are now selected by name from the registry's top-level `normalization` block.

- **Mixed normalization (GLM channels)** — per-channel `glm_mask` / `conditioning_glm_mask` buffers gate between `log1p/expm1` (GLM channels) and standard mean/std normalization, applied in `_forward` and `normalize_conditioning`.

- **Static invariant channels** — `topo` and `nexrad_proximity` arrays are loaded from the package and registered as buffers; appended in `build_condition` (nexrad then topo, broadcast over batch), enabling MRMS+GLM variants.

- **`GOESGLMGrid` data source** — new class in `earth2studio/data/goes_glm.py` wrapping `GOESGLM` event parsing into a gridded 0.1° product via 5-minute binning and `histogram2d`, matching preprocessing that was applied to model training data. Emits `xr.DataArray[time, variable, lat, lon]` with `glm_density` and `glm_energy_density` variables; exported from `earth2studio.data`. Also fixes a session-reuse bug in `GOESGLM._fetch` where the aiobotocore client could be torn down between repeated calls.

- **AMP + `torch.compile` support** — `StormScopeBase` gains `amp` and `compile` constructor arguments. When `amp=True`, DiT forward passes run under `torch.autocast`; sampler latent math stays in fp64. `compile_experts()` wraps each staged expert with `torch.compile(mode="reduce-overhead")`. Roughly 5x speedup for inference

- **Dependency bump** — `nvidia-physicsnemo>=2.1.0` (was `>=2.0`).

- **Tests** — new parametrized tests cover AMP/compile flag propagation, `mrms_coverage_mask` AND-combination with the interpolator mask, and invalid fill-constant behavior for both the default and legacy paths.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.
- [ ] Assess and address Greptile feedback (AI code review bot for guidance; use discretion, addressing all feedback is not required).

### CODE REVIEW

File: earth2studio/models/px/stormscope.py [hunk 1]
```diff
@@ -1201,113 +1703,83 @@ def load_model(
...
+            topo=topo,
+            nexrad_proximity=nexrad_proximity,
             input_times=input_times,
             output_times=output_times,
             y_coords=y,
             x_coords=x,
             input_interp_max_dist_km=6.0 * spatial_downsample,
+            amp=amp,
+            compile=compile,
         )
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:1792]: Addressed in [3df05a0](https://github.com/NVIDIA/earth2studio/pull/938/commits/3df05a01482b4c0495a93e912f59f47f62a6f6db)

File: earth2studio/models/px/stormscope.py [hunk 2]
```diff
@@ -1555,95 +2198,81 @@ def load_model(
...
+        # Static invariant channels (loaded only when the variant requests them)
+        topo = cls._load_invariant(package, "topo.npy", pkg) if pkg.get("topo") else None
+        nexrad_proximity = (
+            cls._load_invariant(package, "nexrad_proximity.npy", pkg)
+            if pkg.get("nexrad_proximity")
+            else None
+        )
 
-        return cls(
+        # NEXRAD circular coverage mask: defines which pixels were valid MRMS
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:2364]: Addressed in [3df05a0](https://github.com/NVIDIA/earth2studio/pull/938/commits/3df05a01482b4c0495a93e912f59f47f62a6f6db)

File: earth2studio/data/goes_glm.py [hunk 3]
```diff
@@ -721,3 +729,160 @@ def _normalize_lat_lon_bbox(
...
+                    f"Variable id {v!r} not supported by GOESGLMGrid. "
+                    f"Available: {list(self._VARIABLE_MAP)}"
+                )
+
+        ny, nx = self._lat_centres.size, self._lon_centres.size
+        out = np.zeros((len(time_list), len(variable_list), ny, nx), dtype=np.float32)
+
+        underlying = sorted({self._VARIABLE_MAP[v] for v in variable_list})
+        for ti, t in enumerate(time_list):
+            df = self._events(t, underlying)
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply earth2studio/data/goes_glm.py]: Addressed in [3df05a0](https://github.com/NVIDIA/earth2studio/pull/938/commits/3df05a01482b4c0495a93e912f59f47f62a6f6db)

**Comment by pzharrington** [comment]
@greptileai

**Comment by pzharrington** [comment]
@greptileai

File: earth2studio/models/px/stormscope.py [hunk 4]
```diff
@@ -1555,95 +2186,83 @@ def load_model(
...
             input_times=input_times,
             output_times=output_times,
             input_interp_max_dist_km=6.0 * spatial_downsample,
             conditioning_interp_max_dist_km=6.0 * spatial_downsample,
+            amp=amp,
+            compile=compile,
         )
+        if pkg.get("deprecated", False):
+            model._INPUT_INVALID_FILL_CONSTANT = cls._LEGACY_INPUT_INVALID_FILL_CONSTANT
+        return model
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:2407]: Not relevant, we don't interact with `state_dict` of model wrappers and checkpoint-resume or distributed training are not done with these inference wrappers. The wrappers are reconstructed by calling `load_model`, which sets the legacy fill constant in the first place, so there's no silent reversion path in practice.

File: earth2studio/data/goes_glm.py [hunk 5]
```diff
@@ -317,8 +318,12 @@ async def fetch(
         pd.DataFrame
             Event-level lightning observations.
         """
-        if self.fs is None:
-            await self._async_init()
+        # Always build a fresh asynchronous filesystem for this fetch. The
```
Comment (NickGeneva) [inline earth2studio/data/goes_glm.py:321]: why was the previous causing issues?

File: earth2studio/data/goes_glm.py [hunk 6]
```diff
@@ -721,3 +729,167 @@ def _normalize_lat_lon_bbox(
                 f"into two boxes. Got {lat_lon_bbox}."
             )
     return (lat_min, lon_min, lat_max, lon_max)
+
+
+@check_optional_dependencies()
+class GOESGLMGrid:
```
Comment (NickGeneva) [inline earth2studio/data/goes_glm.py:735]: Add the the data source rst file

File: earth2studio/data/goes_glm.py [hunk 7]
```diff
@@ -721,3 +729,167 @@ def _normalize_lat_lon_bbox(
...
+    """Gridded GOES GLM lightning product for StormScope.
+
+    Wraps :py:class:`GOESGLM` (a per-event LCFA source) and turns the event
+    point cloud into a regular **0.1-degree** lat/lon grid by 5-minute temporal
+    binning and 2D histogramming, matching the GLM product the StormScope
+    MRMS+GLM nowcast model was trained on. Unlike :py:class:`GOESGLM` (which
+    returns a :py:class:`pandas.DataFrame` of events), this source returns a
+    gridded :py:class:`xarray.DataArray` consumable by
```
Comment (NickGeneva) [inline earth2studio/data/goes_glm.py]: This should be clear since its a data source, little too verbose here maybe

File: earth2studio/data/goes_glm.py [hunk 8]
```diff
@@ -721,3 +729,167 @@ def _normalize_lat_lon_bbox(
...
+    _CONUS_BBOX = (24.5, -125.0, 49.5, -66.0)
+    # E2S variable -> underlying GOESGLM event variable.
+    _VARIABLE_MAP = {"glm_density": "flashc", "glm_energy_density": "flashe"}
+
+    def __init__(
+        self,
+        satellite: str = "east",
+        cache: bool = True,
+        verbose: bool = True,
+        **goes_glm_kwargs: object,
```
Comment (NickGeneva) [inline earth2studio/data/goes_glm.py]: Swap out the kwargs here for explicit parameters. Important for the API docs if anything putting software preferences aside.

File: pyproject.toml
Comment (NickGeneva) [inline pyproject.toml:1]: Update change log 0.17.0 section when the rebase is merged

File: earth2studio/data/goes_glm.py [hunk 9]
```diff
@@ -721,3 +729,167 @@ def _normalize_lat_lon_bbox(
...
+        # Bin edges and centres. arange end padded by a small epsilon so the final
+        # edge is included; centres sit at edge + 0.5 * resolution.
+        self._lat_edges = np.arange(self._LAT_MIN, self._LAT_MAX + 1e-9, self._RES)
+        self._lon_edges = np.arange(self._LON_MIN, self._LON_MAX + 1e-9, self._RES)
+        self._lat_centres = 0.5 * (self._lat_edges[:-1] + self._lat_edges[1:])
+        lon_centres = 0.5 * (self._lon_edges[:-1] + self._lon_edges[1:])
+        # Return longitudes in the Earth2Studio [0, 360) convention.
+        self._lon_centres = (lon_centres + 360.0) % 360.0
+
+    def __call__(
```
Comment (NickGeneva) [inline earth2studio/data/goes_glm.py:823]: The async fetch method need an implementation too here

File: earth2studio/models/px/stormscope.py [hunk 10]
```diff
@@ -235,17 +269,332 @@ def __init__(
...
-                "same_names": True,
-            },
+        """Load the default StormScope package from Hugging Face."""
+        return Package(
+            "hf://nvidia/stormscope-goes-mrms@62f0fd2fa52c3cff67c931daac18cdc0d9f58d2a",
+            cache_options={"cache_storage": Package.default_cache("stormscope")},
+        )
+
+    @staticmethod
+    def _load_registry(package: Package) -> dict[str, Any]:
```
Comment (NickGeneva) [inline earth2studio/models/px/stormscope.py:297]: does this need to be a static method, vs inlined?

File: earth2studio/models/px/stormscope.py [hunk 11]
```diff
@@ -162,13 +170,19 @@ def __init__(
...
+        topo: torch.Tensor | None = None,
+        nexrad_proximity: torch.Tensor | None = None,
         sampler_args: dict[str, Any] | None = {"num_steps": 100, "S_churn": 10},
         input_times: np.ndarray = np.array([np.timedelta64(0, "h")]),
         output_times: np.ndarray = np.array([np.timedelta64(1, "h")]),
         y_coords: np.ndarray | None = None,
         x_coords: np.ndarray | None = None,
         input_interp_max_dist_km: float = 12.0,
         conditioning_interp_max_dist_km: float = 26.0,
+        amp: bool = False,
```
Comment (NickGeneva) [inline earth2studio/models/px/stormscope.py]: Should the default be true since its usable?

File: earth2studio/models/px/stormscope.py [hunk 12]
```diff
@@ -1537,13 +2154,27 @@ def load_model(
...
-            - "6km_10min_natten_pure_obs_mrms_obs_6steps": 6km resolution, 10 minute timestep
+            - ``"3km_10min"``: 3km resolution, 10 minute timestep, MRMS+GLM nowcasting
+            - ``"6km_1hr"``: 6km resolution, 60 minute timestep, MRMS+GLM nearcasting
 
+            Legacy training-style names are accepted as aliases.
+            Default is ``"3km_10min"``.
         conditioning_data_source : DataSource | ForecastSource | None, optional
-            Data source to use for conditioning, by default None.
+            Data source to use for conditioning (GOES), by default None.
+        glm_data_source : DataSource | None, optional
```
Comment (NickGeneva) [inline earth2studio/models/px/stormscope.py:2299]: what happends when the GLM data source is none?

File: earth2studio/models/px/stormscope.py [hunk 13]
```diff
@@ -1537,13 +2154,27 @@ def load_model(
...
+            Gridded GLM source (e.g. :py:class:`earth2studio.data.GOESGLMGrid`) used
```
Comment (NickGeneva) [inline earth2studio/models/px/stormscope.py]: Nit: first time GLM mentioned, can you spell out what it stands for in the API doc for the model

File: earth2studio/models/px/stormscope.py
Comment (NickGeneva) [inline earth2studio/models/px/stormscope.py:268]: So if I understand correctly, variables should not have the glm field added?

Comment (NickGeneva) [inline earth2studio/models/px/stormscope.py:1]: With `load_model` would some one really ever not want to use AMP? 

I could see having compile as an option, but AMP doesnt have the same over head. on the first pass

Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:1]: Potentially for numerical stability/accuracy but it's certainly an edge case

File: earth2studio/data/goes_glm.py (hunk 5 above)
Comment (pzharrington) [reply earth2studio/data/goes_glm.py:321]: There was some bug happening in mutli-step rollouts...honestly don't quite remember the details but this fixed it

File: earth2studio/models/px/stormscope.py [hunk 14]
```diff
@@ -1357,6 +1818,13 @@ class StormScopeMRMS(StormScopeBase):
     conditioning_interp_max_dist_km : float, optional
         Maximum distance in kilometers for nearest neighbor interpolation of conditioning data.
         Points beyond this distance are masked as invalid. Default is 26.0.
+    mrms_coverage_mask : torch.Tensor | None, optional
+        Boolean NEXRAD-coverage mask of shape ``[H, W]`` on the model grid, True
+        where MRMS data is considered valid (inside NEXRAD circular coverage).
+        When provided, it is used as the initial ``valid_mask`` and is ANDed with
+        any interpolator-derived mask built by :meth:`build_input_interpolator`.
+        Loaded automatically from the package for non-deprecated variants.
+        Default is None.
```
Comment (pzharrington) [inline earth2studio/models/px/stormscope.py:1872]: `glm_data_source` needs documenting

File: earth2studio/models/px/stormscope.py (hunk 12 above)
Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:2299]: Nice catch, yeah this fell through the cracks a bit. As of now it will error on the forward pass if the GLM data source is none. Might see if I can refactor to not require it (keep it optional), and in the case of `None` require the caller to pass GLM that's been appropriately preprocessed similar to `call_with_conditioning`

Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:2299]: The GLM preproc I felt was too gnarly and specific to try and fold into the existing `conditioning_data_source` usage and related handling

File: earth2studio/models/px/stormscope.py
Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:268]: Variables do include it. See the other comment for explanation

File: earth2studio/models/px/stormscope.py (hunk 12 above)
Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:2299]: Should be clarified now with some minor implementation tweaks. Best described by this note which is added to the `StormScopeMRMS` docstring:

    **GLM state channel.**  The ``3km_10min`` variant includes a ``glm_density``
    channel (gridded GLM lightning counts, normalized with ``log1p``) as part of
    its *state* — both an input observation and a predicted output (the
    ``6km_1hr`` variant has no GLM channel).  Because the GLM source lives on a
    different native grid from MRMS, ``glm_density`` is handled separately from
    the radar channels and is the GLM analogue of the GOES ``conditioning``:

    * **Auto path** (:meth:`__call__` / :meth:`~StormScopeBase.create_iterator`):
      pass ``glm_data_source`` (e.g. :py:class:`earth2studio.data.GOESGLMGrid`)
      to ``load_model`` and GLM is fetched, bilinearly regridded, and injected
      into the state automatically on every step — exactly as
      ``conditioning_data_source`` is fetched via :meth:`fetch_conditioning`.
      The GLM bilinear interpolator is built lazily on the first call. The input
      state ``x`` only needs its radar channels populated (the GLM channels are
      overwritten); a zero placeholder is fine. In this case, the model will be
      using ground-truth GLM observations during the rollout, so is not doing
      pure forecasting (and can only be run for dates in the past where the full
      timeseries of GLM observations is available).

    * **Coupled path** (:meth:`~StormScopeBase.call_with_conditioning`): just as
      this method takes ``conditioning`` from the caller rather than the data
      source, it leaves the *entire* state — GLM included — to the caller and
      never touches ``glm_data_source``.  Populate the GLM channels of ``x``
      yourself (e.g. via :meth:`fetch_glm` for the initial state); during the
      rollout GLM then flows autoregressively from the model's own predictions,
      like the radar channels. This is the more typical pure-forecast use case.

File: earth2studio/models/px/stormscope.py (hunk 10 above)
Comment (pzharrington) [reply earth2studio/models/px/stormscope.py:297]: It's called in 4 places across the class, so keeping it as a static method avoids repeating the open/json.load pattern. Happy to inline if you still prefer, but felt the deduplication was worth it.

File: pyproject.toml
Comment (pzharrington) [reply pyproject.toml:1]: Done

File: earth2studio/models/px/stormscope.py [hunk 15]
```diff
@@ -1523,12 +2253,31 @@ def prep_input(
...
+        if self.n_glm_channels > 0 and self.glm_data_source is not None:
+            return self._inject_glm(x, coords)
+        return x
+
     @classmethod
     def load_model(
         cls,
         package: Package,
-        model_name: str = "6km_60min_natten_cos_zenith_input_mrms_eoe",
+        model_name: str = "3km_10min",
```
Comment (NickGeneva) [inline earth2studio/models/px/stormscope.py]: Maybe this should be a literal[ "3km_10min", ...] so it self documents the options in the type

**Review by NickGeneva — APPROVED** [review]
one last comment, approving!

File: earth2studio/models/px/stormscope.py (hunk 15 above)
Comment (pzharrington) [reply earth2studio/models/px/stormscope.py]: Nice, added
